File: backend/src/main/java/com/example/contact/domain/valueobject/PhoneNumber.java
package com.example.contact.domain.valueobject;

import com.example.contact.domain.exception.InvalidContactDataException;
import java.util.regex.Pattern;

/**
 * Optional phone number normalized to digits and leading plus sign.
 */
public final class PhoneNumber {

    private static final Pattern ALLOWED = Pattern.compile("^[+0-9]{7,30}$");

    private final String value;

    private PhoneNumber(String value) {
        this.value = value;
    }

    /**
     * Creates a phone number from raw input.
     *
     * @param raw user input; blank yields empty optional
     * @return phone number or null when absent
     * @throws InvalidContactDataException when format is invalid
     */
    public static PhoneNumber createOptional(String raw) {
        if (raw == null || raw.isBlank()) {
            return null;
        }
        String normalized = raw.trim().replaceAll("[\\s-]", "");
        if (!ALLOWED.matcher(normalized).matches()) {
            throw new InvalidContactDataException("Invalid phone number format");
        }
        return new PhoneNumber(normalized);
    }

    /**
     * Rehydrates from persistence.
     *
     * @param stored stored phone
     * @return phone instance
     */
    public static PhoneNumber fromStored(String stored) {
        return new PhoneNumber(stored);
    }

    public String value() {
        return value;
    }

    @Override
    public String toString() {
        return value;
    }
}

File: frontend/src/contact/domain/valueobject/PhoneNumber.ts
import { InvalidContactDataError } from '../exception/InvalidContactDataError';

const PHONE_PATTERN = /^[+0-9]{7,30}$/;

/**
 * Optional phone number normalized to digits and leading plus.
 */
export class PhoneNumber {
  private constructor(private readonly value: string) {}

  /**
   * Creates a phone number from raw input.
   * @param raw - user input; blank yields null
   * @returns phone or null when absent
   * @throws InvalidContactDataError when format is invalid
   */
  static createOptional(raw: string | null | undefined): PhoneNumber | null {
    if (!raw?.trim()) {
      return null;
    }
    const normalized = raw.trim().replace(/[\s-]/g, '');
    if (!PHONE_PATTERN.test(normalized)) {
      throw new InvalidContactDataError('Invalid phone number format');
    }
    return new PhoneNumber(normalized);
  }

  /**
   * Rehydrates from stored value.
   * @param stored - persisted phone
   * @returns phone instance
   */
  static fromStored(stored: string): PhoneNumber {
    return new PhoneNumber(stored);
  }

  /**
   * Returns the phone string value.
   * @returns phone digits
   */
  toString(): string {
    return this.value;
  }
}
